src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -0,0 +1,921 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.lang.runtime;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.MethodHandles.Lookup.ClassOption;
+import java.lang.invoke.MethodType;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import jdk.internal.misc.VM;
+import jdk.internal.org.objectweb.asm.ClassWriter;
+import jdk.internal.org.objectweb.asm.FieldVisitor;
+import jdk.internal.org.objectweb.asm.MethodVisitor;
+import jdk.internal.org.objectweb.asm.Type;
+
+import static jdk.internal.org.objectweb.asm.Opcodes.*;
+
+/**
+ * This  class is used to create objects that have number and types of
```

Comment:
Agreed.  Also, this class is unusual in that it is not instantiated; like `Arrays` or `Collections` it is a (small) bundle of static factories (or are they algorithms?).  I think as such it should be called `Carriers`.

I'm slightly surprised the MH factories are not factored through a metaobject of the form
```
record CarrierBinding(
    MethodType methodType,
    MethodHandle constructor,
    List<Class<?>> componentTypes,
    List<MethodHandle> components)
{ … }
```

The presupposition here, I suppose, is that carriers will only be used by condy or similar quasi-statically configured clients, so having the multiple lookups through a hidden table is no burden, and the clients can always keep the associations correct (between constructors and various component accessors).

**But** if I were to use carriers to manage intermediate structures in (say) a serialization package (for instance) I would need to make my own records like the above for my own bookkeeping, and I would be slightly miffed that the Carrier API insisted on doing a de-novo lookup twice on each MT key (to say nothing of me having to create the MT key first).

**And** if I were to use carriers in that way (away from condy), **then** I would want to skip the step of building the MethodType, and wish for a factory method for CarrierBinding instances that took a plain List<Class>, as well as a factory method that took the MethodType (which is convenient for condy).

BTW, it would be normal to give the name `Carrier` (which is a good name BTW) to the record type that embodies the group of method handles, so `record Carrier(…constructor…components…) {…factories…}`.

I suppose stuff like this could be added later.  But it's worth considering now, simply because there is an early decision point between a class named `Carrier` and a static-only class named `Carriers`.

Reply:
Will do. When I wrote this code, I wasn't expecting the external exposure. Hence...